### ISSUE TITLE
Azure Service Bus Processors - MaxReceiveWait

### ISSUE THREAD

**Issue by davidallyoung** [Azure/azure-sdk-for-net#16773]
**Query/Question**
How can we help?
Hi, I recently had a short conversation with @JoshLove-msft on #15931 regarding the removal of the MaxReceiveWait. He suggested I open up an issue for gathering more feedback from the community on this property's usage.

My team and I after experimenting without it for a few days have actually found it would be handy to have the property available when using the processors so we can control the wait time between session messages independently of the other operations that seems to also be controlled by the ServiceBusRetryOptions.TryTimeout. Let me know if I should have created this as a feature request instead of as a Question/Query and I can make a new one. Thanks again to @JoshLove-msft for opening up conversation on this!

**Environment:**
 - Name and version of the Library package used: [e.g. Azure.Storage.Blobs 12.2.0] 
Azure.Messaging.ServiceBus 7.0.0-preview-9
 - Hosting platform or OS and .NET runtime version (`dotnet --info` output for .NET Core projects): [e.g. Azure AppService or Windows 10 .NET Framework 4.8]
.NET Core 3.1, hosted in various locations, Azure App Service and containerized instances.
 - IDE and version : [e.g. Visual Studio 16.3]
Visual Studio 16.7.7

**Comment by jsquire** [comment]
//cc: @ramya-rao-a, @richardpark-msft, @MiYanni, @HarshaNalluru, @hemanttanwar, @YijunXieMS, @srikanta, @KieranBrantnerMagee

**Comment by jsquire** [comment]
//fyi: @SeanFeldman, @danielmarbach

**Comment by jsquire** [comment]
Thanks for opening this, @davidallyoung.  Tagging and routing for discussion.

**Comment by JoshLove-msft** [comment]
Thanks @davidallyoung - I think the use case you mentioned in https://github.com/Azure/azure-sdk-for-net/pull/15931 makes sense:
 > In our specific case, we have many Sessions where each session has pretty sparse message density, or rather a low volume over short periods of time, so we like to bail out well before a minute (which is the default by my understanding) has passed and start picking up other sessions that have work.

To help prioritize this, can I ask if this issue is blocking you from using the new library?

**Comment by davidallyoung** [comment]
We are still experimenting with it for the next few days so it's hard to say, but I don't think it will block us in the near term (next month or two). As we deploy some workloads to it we may find that the extra knob would be helpful for us and I can report here on that if it comes up.

**Comment by JoshLove-msft** [comment]
> We are still experimenting with it for the next few days so it's hard to say, but I don't think it will block us in the near term (next month or two). As we deploy some workloads to it we may find that the extra knob would be helpful for us and I can report here on that if it comes up.

Thanks @davidallyoung - sounds great!

**Comment by danielmarbach** [comment]
The originally reference discussion said 

> In our specific case, we have many Sessions where each session has pretty sparse message density, or rather a low volume over short periods of time, so we like to bail out well before a minute (which is the default by my understanding) has passed and start picking up other sessions that have work.

Isn't such a bail-out scenario effectively an advanced session usage scenario and thus you wouldn't want to use the processor as a higher-level abstraction but actually take control of session management entirely yourself?

Maybe the important question to ask here is not "Can we reintroduce the property" but what is the intent of the processor and to what extent should it be configurable and when do advanced scenario start that requires a user to leave the abstraction land and go to work directly with sessions? As far as I understand the scenario of "bail" out could be achieved by interacting with a SessionReceiver directly and dispose of it when deemed necessary?

**Comment by JoshLove-msft** [comment]
> Maybe the important question to ask here is not "Can we reintroduce the property" but what is the intent of the processor and to what extent should it be configurable and when do advanced scenario start that requires a user to leave the abstraction land and go to work directly with sessions? As far as I understand the scenario of "bail" out could be achieved by interacting with a SessionReceiver directly and dispose of it when deemed necessary?

It's a good point. Caring about the time each receive call will wait is certainly a more advanced scenario, but I wouldn't think this automatically means you should not be using the processor. I think there is a tradeoff between the scenarios would be enabling and the added cognitive overhead of introducing a somewhat complex property that most users will not care about. I think it will be great to see how users are using the processor in the wild to help inform the decision as to whether the property should be reintroduced.

**Comment by davidallyoung** [comment]
I definitely agree we're probably in a more advanced scenario, however as far as I can tell at the moment in our experimenting with it, we would probably re-implement the session processor almost exactly just to expose the receive time. We've only been tinkering for a few days though, I could definitely be off on that assumption.

In context of others migrating to this new SDK though, someone coming from the previous SDK may be expecting to find an equivalent to the SessionHandlerOptions.MessageWaitTimeout, which feels similar to what the MaxReceiveWait time would have done in this SDK. Ref: https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Microsoft.Azure.ServiceBus/src/SessionHandlerOptions.cs#L64

**Comment by ErikPilsits-RJW** [comment]
I would also like to see this option return, as well as access to the session. In the old SDK the handler had access to the underlying session.
```csharp
client.RegisterSessionHandler(
                        async (session, message, cancellationToken) =>
```
This gave us the ability to allow for some different options. We could set a max number of messages received per session before we would manually close the session.

`session.CloseAsync()`

In our use case, we have session enabled topic subscriptions to which interleaved order data is submitted. The session ID is usually set to the order ID. The purpose for sessions is to support scaled app instances, because it is imperative that messages related to a particular entity, in this case an order ID, are processed in order. But it is not a requirement that individual orders are processed in any order, nor that all messages relating to an order are processed at once (more events can be added at any time as modifications are made).

We are sync'ing data between systems, an integration layer, so it's also important that it happens quickly. The current implementation is to process only 1 message at a time from any given session, then close the session. This guarantees that distributed instances can't process a particular order's messages in parallel, and also that we're giving fairness to many orders entering the system at once and processing them in generally the order they were received.

With the new SDK this doesn't seem possible with the Processor. Even if we had MaxReceiveWait, I can't enforce processing only 1 session message at a time without the ability to close the session (right now that only happens after a receive timeout). I really need both, so that if our implementation were to change, and I want to process 5 messages from an order, I can at least timeout reasonably if there are only 3 messages available. We have to keep the number of concurrent session listeners to about 4 so that we don't crush our downstream apis, so binding up a listener waiting for a message for 60 or even 15 seconds would really hurt system performance. And I don't see any other good way to throttle downstream calls without risking losing a message / session lock.

Trying to use TryTimeout at a low value isn't a great solution because it also affects the accept session call, and there's no reason to override the default 60 seconds on that. (A solution for token Cancellation would be great to help app shutdown cleanly there also.)

I referenced several github issues that I read today which relate to our current implementation with the old SDK, so I guess the point is right now I can't replicate what we're doing which means we can't use the Processor from the new SDK at this time. And considering we use the base functionality that the handlers / processors provide, it's not something I'm keen to reimplement.

**Comment by JoshLove-msft** [comment]
Thanks for the feedback @ErikPilsits-RJW - we have several recorded scenarios where this option would be helpful at this point, so I think we are in a spot where we can investigate adding this option.

**Comment by ErikPilsits-RJW** [comment]
Thank you @JoshLove-msft . Does that investigation also include access to the session as I described? I think the concepts go hand in hand, but I can open a new feature to tie to this one if it's better tracked like that.

**Comment by JoshLove-msft** [comment]
>  I can't enforce processing only 1 session message at a time without the ability to close the session (right now that only happens after a receive timeout). 

I'm not sure I follow this line - if you have MaxCallsPerSession = 1, then it would only process 1 message per session at a time. 

Can you elaborate on the logic you would use to determine when to move on to the next session? I would expect that having a low MaxReceiveWait would keep the processor busy as it would switch to new sessions quickly.

**Comment by epilsits** [comment]
Still me, different account.

So assume I have 50 messages per session and 50 sessions (orders). Events come in fast, for example the order header record and then line items then other data fields, but they are all interleaved; all header records might be first, then mixed in line items and data. It's important to get the orders created in the destination system, not necessarily get all the line items created. Another scenario is grouping those orders together, which benefits having the orders created first before moving on to the grouping events.

I always do 1 call per session because session events must be processed in chronological order. But no matter what I do with wait times, the Processor would always process the 50 messages associated with that order (order id = session id), which is not desirable.

So we currently close the session after every message, allowing the handler to get the next available session. I realize that's not ideal, so I'm working on implementing a max messages per session that will track the number of messages received per session and close the session when it is reached. The hope is to get benefits of both processing several messages without the overhead of close/open sessions, and to be fair to the data coming in.

**Comment by JoshLove-msft** [comment]
Got it, thanks for the explanation. We can investigate this as well but I'd rather keep this as a separate issue than the MaxReceiveTime.

**Comment by epilsits** [comment]
I can create a new feature issue tomorrow, reference this discussion.

**Comment by nuzolx** [comment]
@epilsits Did you create it? we are facing the same issue.
